builder/slide_and_fold_storage.cc

```diff
@@ -354,13 +405,35 @@ class slide_and_fold : public node_mutator {
     merge_crop(bounds, op->bounds);
     if (bounds) {
       substitute_bounds(*bounds, current_buffer_bounds());
+      // Now do the reverse substitution, because the updated bounds can be used in other
+      // bounds.
+      // NOTE(vksnk): I'm not sure this needed anymore, but seems logical to update it in
+      // both directions.
+      substitute_bounds(current_buffer_bounds(), op->sym, *bounds);
+
       // This simplify can be heavy, but is really useful in reducing the size of the
       // expressions.
       for (auto& b : *bounds) {
         b = simplify(b);
       }
     }
+
     auto set_bounds = set_value_in_scope(current_buffer_bounds(), op->sym, bounds);
+
+    // We only want to fold if we are inside of the loop and the cropped buffer
```

Comment:
This could be moved to a helper and called from both crop_buffer and crop_dim, I think it should be. It's weird for crop_buffer and crop_dim to be treated differently.

In general I have attempted to simplify before slide_and_fold_storage, I'd like to try again after this change lands. This is an obvious thing that would break.

Reply:
Sure, I changed it to be called from both crop_buffer and crop_dim. I am wondering if we should skip crop_buffer during construction if there is a crop_dim for that buffer already in the specific loop.